Guides 2.0 Actions # parameters (#121)
hanami/guides: content/v2.0/actions/parameters.md

---
title: Parameters
order: 30
---

Parameters are taken from the Rack env, wrapped into the `Params` object, and passed as an argument to `#handle` method.

They are similar to a Ruby `Hash`, but they offer an expanded set of features.

## Sources

Params can come from:

  * [Router variables](/v2.0/router/basic-usage) (eg. `/books/:id`)
  * Query string (eg. `/books?title=Hanami`)
  * Request body (eg. a `POST` request to `/books`)

## Access

To access the value of a param, we can use the _subscriber operator_ `#[]`.

```ruby
# app/controllers/dashboard/index.rb

module Bookshelf
  module Actions
    module Books
      class Index < Action

>         def handle(req, res)

Reviewer: We're using `request`, and `response` after all so it should be used here as well.

Author: Addressed.

          req.body = "Query string: #{ req.params[:q] }"
        end
      end
    end
  end
end
```

If we visit `/dashboard?q=foo`, we should see `Query string: foo`.

### Symbol Access

Params and nested params can be referenced **only** via symbols.

```ruby
params[:q]
params[:book][:title]
```

Now, what happens if the parameter `:book` is missing from the request?
Because `params[:book]` is `nil`, we can't access `:title`.
In this case Ruby will raise a `NoMethodError`.

We have a safe solution for our problem: `#get`.
It accepts a list of symbols, where each symbol represents a level in our nested structure.

```ruby
params.get(:book, :title)             # => "Hanami"
params.get(:unknown, :nested, :param) # => nil instead of NoMethodError

> ```

Reviewer: Oh noez, I had no idea this exists. Too late now but...maybe not? If we also provide `dig` then please do change it to `dig` here, if not...then oh well, something we can address in 3.0.

Author: Both work the same - replaced with `dig`. We can remove the `get` if no needed in the upcoming releases.

## Whitelisting

In order to show how whitelisting works, let's create a new action: `app/actions/signup/create`

We want to provide self-registration for our users.
We build a HTML form which posts to an action that accepts the payload and stores it in the `users` table.
That table has a boolean column `admin` to indicate whether a person has administration permissions.

A malicious user can exploit this scenario by sending this extra parameter to our application, thereby making themselves an administrator.

We can easily fix this problem by filtering the allowed parameters that are permitted inside our application.
Please always remember that **params represent untrusted input**.

We use `.params` to map the structure of the (nested) parameters.

```ruby
# app/actions/signup/create.rb

module Bookshelf
  module Actions
    module Signup
      class Create < Action
        params do
          required(:email).filled(:str?)
          required(:password).filled(:str?)
    
          required(:address).schema do
            required(:country).filled(:str?)

>           end

Reviewer: Please don't use type predicates and switch to type specs. So `:str?` becomes `:string` etc. This is how it's going to work in dry-schema 2.0 by default and it will be a mandatory argument.

Author: done

        end
    
        def handle(req, res)
          puts req.params[:email]             # => "alice@example.org"
          puts req.params[:password]          # => "secret"
          puts req.params[:address][:country] # => "Italy"
    
          puts req.params[:admin]             # => nil
        end
      end
    end
  end
end
```

Even if `admin` is sent inside the body of the request, it isn't accessible from `params`.

## Validations & Coercion

### Use Cases

In our example (called _"Signup"_), we want to make `password` a required param.

Imagine we introduce a second feature: _"Invitations"_.
An existing user can ask someone to join.
Because the invitee will decide a password later on, we want to persist that `User` record without that value.

If we put `password` validation in `User`, we need to handle these two use cases with a conditional.
But in the long term this approach is painful from a maintenance perspective.

```ruby

> # Example of poor style for validations

Reviewer: I would drop this comment. In general we should avoid presenting other solutions in a negative way. I used to do this a lot thinking that I'm just explaining why I do X when something else does Y but it actually makes things worse as people may get upset when what they enjoy is being presented as "worse" or "poor style" etc.

It should be sufficient to explain why certain things work the way they work and the reasoning behind it without contrasting it with other solutions so often.

Author: I agree. Official docs should NOT be subjective. I can present things filled in with more personal opinion in un/semi-official channels like Hanami Mastery.

class User
  attribute :password, presence: { if: :password_required? }

  private
  
  def password_required?
     !invited_user? && !admin_password_reset?
  end
end
```

We can see validations as the set of rules for data correctness that we want for **a specific use case**.

> For us, a `User` can be persisted with or without a password, **depending on the workflow** and the route through

Reviewer: ```suggestion
For us, a `User` can be persisted with or without a password, **depending on the use case** and the route through
```

I think it's fine to repeat the same word here as `workflow` has specific meanings in various contexts (ie background job workflow etc.).

which the `User` is persisted.

### Boundaries

The second important aspect is that we use validations to prevent invalid inputs to propagate in our system.
In an MVC architecture, the model layer is the **farthest** from the input.
It's expensive to check the data right before we create a record in the database.

> If we **consider correct data as a precondition** before starting our workflow, we should stop unacceptable inputs as soon as possible.

Reviewer: ```suggestion
If we **consider correct data as a precondition** before starting our workflow, we should stop invalid inputs as soon as possible.
```

Think of the following method.
We don't want to continue if the data is invalid.

```ruby
def expensive_computation(argument)
  return if argument.nil?
  # ...
end
```

### Usage

We can coerce the Ruby type, validate if a param is required, determine if it is within a range of values, etc..

```ruby
# app/actions/signup/create.rb
module Bookshelf
  module Actions
    module Signup
      class Create < Action
        MEGABYTE = 1024 ** 2

        params do
          required(:name).filled(:str?)
          required(:email).filled(:str?, format?: /@/)
          required(:password).filled(:str?)
          required(:terms_of_service).filled(:bool?)
          required(:age).filled(:int?, included_in?: 18..99)
          optional(:avatar).filled(size?: 1..(MEGABYTE * 3))

>         end

Reviewer: All of this needs to be updated to use type specs, `:str?` -> `:string` `:bool?` -> `:bool` `:int?` -> `:integer` and avatar should be `:string` before the size? check

Author: Done. Replaced all.

        def handle(req, res)
          if req.params.valid?
            # ...
          else
            # ...
          end
        end
      end
    end
  end
end
```

Parameter validations are delegated, under the hood, to [Hanami::Validations](https://github.com/hanami/validations).
Please check the related documentation for a complete list of options and how to share code between validations.

## Concrete Classes

> The params DSL is really quick and intuitive but it has the drawback that it can be visually noisy and makes it hard to unit test.

Reviewer: This could be explained better. ie Why does it make things harder to unit test? I would also drop "unit test" and just say "test".

This validation serves several purposes, including allowlisting (ensuring that only allowable params are extracted from a request) and coercion (converting string parameters to boolean, integer, time and other types).

Let's take a look at a books index action that accepts two parameters, `page` and `per_page`.

```ruby
# app/actions/books/index.rb

# frozen_string_literal: true

module Bookshelf
  module Actions
    module Books
      class Index < Bookshelf::Action
        params do
          optional(:page).value(:integer)
          optional(:per_page).value(:integer)
        end

        def handle(request, response)
          request.params[:page]
          request.params[:per_page]
        end
      end
    end
  end
end
```

The schema in the params block specifies the following:

- `page` and `per_page` are both optional parameters
- if `page` is present, it must be an integer
- if `per_page` is present, it must be an integer

With this schema in place, a request with a query string of `/books?page=1&per_page=10` will result in:

```ruby
request.params[:page] # => 1
request.params[:per_page] # => 10
```

> Note that, due to the `params` block, 1 and 10 have been coerced to integer values (with this they would otherwise be strings).

Reviewer: ```suggestion
Notice that thanks to the defined params schema with types, `"1"` and `"10"` are coerced to their integer representations `1` and `10`.
```

> Additional rules can be added to apply further contraints. The following params block specifies that, when present, `page` and `per_page` be greater than or equal to 1, and also that `per_page` be less than or equal to 100.

Reviewer: ```suggestion
Additional rules can be added to apply further constraints. The following params block specifies that, when present, `page` and `per_page` must be greater than or equal to 1, and also that `per_page` must be less than or equal to 100.
```

```ruby
params do
  optional(:page).value(:integer, gte?: 1)
  optional(:per_page).value(:integer, gte?: 1, lteq?: 100)
end
```

> Importantly, now that our params block is doing more than just type coercion, we need to explicitly check for and handle our parameters being invalid.

Reviewer: ```suggestion
Importantly, now that our params schema is doing more than just type coercion, we need to explicitly check for and handle our parameters being invalid.
```

The `#valid?` method on the params allows the action to check the parameters in order halt and return a `422 Unprocessable` response.

```ruby
# app/actions/books/index.rb

# frozen_string_literal: true

module Bookshelf
  module Actions
    module Books
      class Index < Bookshelf::Action
        params do
          optional(:page).value(:integer, gte?: 1)
          optional(:per_page).value(:integer, gte?: 1, lteq?: 100)
        end

        def handle(request, response)
          halt 422 unless request.params.valid?

          # At this point, we know the params are valid
          request.params[:page]
          request.params[:per_page]
        end
      end
    end
  end
end
```

Here's a further example, this time for an action to create a user.

```ruby
# app/actions/users/create.rb

# frozen_string_literal: true

module Bookshelf
  module Actions
    module Users
      class Create < Bookshelf::Action
        params do
          required(:email).filled(:string)
          required(:password).filled(:string)

>           required(:address).schema do

Reviewer: ```suggestion
          required(:address).hash do
```

plain `schema` doesn't specify the type to be `:hash`, which is what `hash` macro does.

            required(:street).filled(:string)
            required(:country).filled(:string)
          end
        end

        def handle(request, response)
          halt 422 unless request.params.valid?

          request.params[:email]             # => "alice@example.org"
          request.params[:password]          # => "secret"
          request.params[:address][:country] # => "Italy"

          request.params[:admin]             # => nil
        end
      end
    end
  end
end
```

The `params` block in this action specifies that:

- `email`, `password` and `address` parameters are required to be present.
- `address` has `street` and `country` as nested parameters, which are also required.

> - each of `email`, `password`, `street` and `country` must be filled (non-blank) strings.

Reviewer: ```suggestion
- `email`, `password`, `street` and `country` must be filled (non-blank) strings.
```

The errors associated with a failed parameter validation are available via `request.params.errors`.

Assuming that the users create action was part of a JSON API, we could render these errors by passing a body when calling `halt`:

```ruby
halt 422, {errors: request.params.errors}.to_json unless request.params.valid?
```

For an empty `POST` request with an empty address object, this action would render:

```json
{
    "errors": {
        "email": [
            "is missing"
        ],
        "password": [
            "is missing"
        ],
        "address": {
            "street": [
                "is missing"
            ],
            "country": [
                "is missing"
            ]
        }
    }
}
```

Action validations use the [dry-validation](https://dry-rb.org/gems/dry-validation/) gem, which provides a powerful DSL for defining schemas.

Consult the [dry-validation](https://dry-rb.org/gems/dry-validation/) and [dry-schema](https://dry-rb.org/gems/dry-schema/) gems for further documentation.

## Using concrete classes

> In addition to specifying parameter validations "inline" in a `params` block, actions can also hand their validation responsibilities to a separate class.

Reviewer: ```suggestion
In addition to specifying parameter validations "inline" in a `params` block, actions can also hand over their validation responsibilities to a separate class.
```

This makes action validations reusable and easier to test independently of the action.

For example:

```ruby
# app/actions/users/params/create.rb

module Bookshelf
  module Actions
    module Users
      module Params
        class Create < Hanami::Action::Params
          params do
            required(:email).filled(:string)
            required(:password).filled(:string)

>             required(:address).schema do

Reviewer: ```suggestion
            required(:address).hash do
```

              required(:street).filled(:string)
              required(:country).filled(:string)
            end
          end
        end
      end
    end
  end
end
```

```ruby
# app/actions/users/create.rb

module Bookshelf
  module Actions
    module Users
      class Create < Bookshelf::Action
        params Params::Create

        def handle(request, response)
          # ...
        end
      end
    end
  end
end
```

## Validations at the HTTP layer

Validating parameters in actions is useful for validating parameter structure, performing parameter coercion and type validations.

More complex domain-specific validations, or validations concerned with things such as uniqueness, however, are usually better performed at layers deeper than your HTTP actions.

> For example, verifying a that an email address has been provided is something an action parameter validation should reasonably do, but checking that a user with that email address doen't already exist is unlikely to be a good responsibility for a HTTP action to have. That validation might instead be peformed by a create user operation, which can perform a check against a user store.

Reviewer: ```suggestion
For example, verifying that an email address has been provided is something an action parameter validation should reasonably do, but checking that a user with that email address doesn't already exist is unlikely to be a good responsibility for an HTTP action to have. That validation might instead be performed by a create user operation, which can perform a check against a user store.
```
